src/Writers/Writers.jl

```diff
@@ -18,23 +18,26 @@ import ..Documenter:
 
 using Compat
 
-# Driver method for document rendering.
-# -------------------------------------
+"""
+A parametric type that allows us to use multiple dispatch to pick the appropriate
+writer for each output format.
 
+The parameter `f` should be an instance of the [`Formats.Format`](@ref) enumeration.
 """
-Writes a [`Documents.Document`](@ref) object to `build` directory in specified file format.
+immutable Writer{f} end
+Writer(f::Formats.Format) = Writer{f}()
+
 """
-function render(doc::Documents.Document)
-    mime = Formats.mimetype(doc.user.format)
-    for (src, page) in doc.internal.pages
-        open(Formats.extension(doc.user.format, page.build), "w") do io
-            for elem in page.elements
-                node = page.mapping[elem]
-                render(io, mime, node, page, doc)
-            end
-        end
-    end
-end
+Writes a [`Documents.Document`](@ref) object to `.user.build` directory in
+the format specified in `.user.format`.
+
+The method should be overloaded in each writer as
+
+    render(::Writer{format}, doc)
+
+where `format` is one of the values of the [`Formats.Format`](@ref) enumeration.
+"""
+render(doc::Documents.Document) = render(Writer(doc.user.format), doc)
```

Comment:
If we can avoid this type instability here that would be great since `Writer(...)` will always be unstable. I doubt it would ever be a performance hotspot, but would be good to avoid anyway.

Here one possible approach:

Don't extend `Writers.render` in each submodule. Just define a new `MarkdownWriter.render` etc. I don't think we'd ever need to dispatch to methods between different modules so including `Writer{Format}` in the signatures doesn't seem worth it. Instead of the dynamic dispatch being used here, we just hardcode the required `if`s:

```julia
function render(doc::Documents.Document)
    doc.user.format === Formats.Markdown ? MarkdownWriter.render(doc) :
    doc.user.format === Formats.HTML     ?     HTMLWriter.render(doc) :
    doc.user.format === Formats.LaTeX    ?    LaTeXWriter.render(doc) : error("unhandled format.")
end
```

I don't see us allowing the output formats to be user-defined any time in the *near* future.

---

Or perhaps write a new `Selector` for this:

```julia
abstract WriterFormat <: Selectors.AbstractSelector

abstract MarkdownFormat <: WriterFormat
abstract HTMLFormat <: WriterFormat
abstract LaTeXFormat <: WriterFormat

Selectors.order(::Type{MarkdownFormat}) = 1.0
Selectors.order(::Type{HTMLFormat})     = 2.0
Selectors.order(::Type{LaTeXFormat})    = 3.0

Selectors.matcher(::Type{MarkdownFormat}, doc) = doc.user.format === Formats.Markdown
Selectors.matcher(::Type{HTMLFormat}, doc)     = doc.user.format === Formats.HTML
Selectors.matcher(::Type{LaTeXFormat}, doc)    = doc.user.format === Formats.LaTeX

Selectors.runner(::Type{MarkdownFormat}, doc) = MarkdownWriter.render(doc)
Selectors.runner(::Type{HTMLFormat}, doc)     = HTMLWriter.render(doc)
Selectors.runner(::Type{LaTeXFormat}, doc)    = LaTeXWriter.render(doc)

render(doc::Documents.Document) = Selectors.dispatch(WriterFormat, doc)
```

which would be user-extensible *if* we choose to export/document output formats at some point.